### PR TITLE
🐛 Implement waitForHead and change waitForBody to use documentReady

### PR DESCRIPTION
**Pull request by cjdeleon62**
Implement waitForHead and change waitForBody to wait for documentReady. This change is for preventing bugs from occurring when using the waitForBody function and allows the body and it's children to render before the Promise resolves. Another method, waitForHead was implemented for cases where waiting for DOMReady would take too long for certain actions.

Resolves: https://github.com/ampproject/amphtml/issues/20377

### CODE REVIEW

**Comment by torch2424** [comment]
Triaging to @jridgewell , Feel free to re-assign 😄

File: src/dom.js [hunk 1]
```diff
@@ -33,11 +34,11 @@ const HTML_ESCAPE_REGEX = /(&|<|>|"|'|`)/g;
 
 /** @const {string} */
 export const UPGRADE_TO_CUSTOMELEMENT_PROMISE =
-    '__AMP_UPG_PRM';
```
Comment (jridgewell) [inline src/dom.js]: These whitespace changes need to be reverted.

Comment (cjdeleon62) [reply src/dom.js]: @jridgewell sorry, I left my Prettier formatter on 😅

File: src/dom.js [hunk 2]
```diff
@@ -87,14 +88,36 @@ export function waitForChildPromise(parent, checkFunc) {
 }
 
 /**
- * Waits for document's body to be available.
+ * Waits for document's head to be available.
+ * @param {!Document} doc
+ * @param {function()} callback
+ */
+export function waitForHead(doc, callback) {
+  waitForChild(doc.documentElement, () => !!doc.head, callback);
```
Comment (jridgewell) [inline src/dom.js]: We should be waiting (ironically) for `doc.body` here, not `doc.head`. The presence of `doc.body` means that the `<head>` (and all its descendants) have been parsed. Waiting only for `doc.head` means the `<head>` tag's opening tag has been seen, but the descendants may not be fully parsed yet.

File: src/dom.js [hunk 3]
```diff
@@ -87,14 +88,36 @@ export function waitForChildPromise(parent, checkFunc) {
...
+ * Waits for document's body to be available and ready.
  * Will be deprecated soon; use {@link AmpDoc#whenBodyAvailable} or
  * @{link DocumentState#onBodyAvailable} instead.
  * @param {!Document} doc
  * @param {function()} callback
  */
 export function waitForBody(doc, callback) {
-  waitForChild(doc.documentElement, () => !!doc.body, callback);
+  waitForChild(doc.documentElement, () => (
+    isDocumentReady(doc) && !!doc.body
```
Comment (jridgewell) [inline src/dom.js]: Ok, it appears this is failing tests for some obscure reasons (MutationObserver inside `waitForChild` doesn't fire during parsing). The fix is to just call `onDocumentReady` from `src/document-ready.js`. We can also change `waitForBodyPromise` to just call `whenDocumentReady`.

Comment (cjdeleon62) [reply src/dom.js]: oh okay, will use those functions 👍

**Comment by cjdeleon62** [comment]
@jridgewell I am unsure if this is what you meant for using `onDocumentReady` inside of the `waitForBody` function?

File: src/dom.js [hunk 4]
```diff
@@ -116,7 +116,7 @@ export function waitForHeadPromise(doc) {
  */
 export function waitForBody(doc, callback) {
   waitForChild(doc.documentElement, () => (
-    isDocumentReady(doc) && !!doc.body
+    onDocumentReady(doc)
```
Comment (jridgewell) [inline src/dom.js]: Sorry, `waitForBody` just calls `onDocumentReady`. There's essentially aliases at this point:

```js
export function waitForBody(doc, callback) {
  return onDocumentReady(doc, callback);
}
```

Same for `waitForBodyPromise`.

Comment (cjdeleon62) [reply src/dom.js]: I think my most recent commit has that change, but tests still seem to be failing for some reason

**Comment by jridgewell** [comment]
Agh, I'm sorry the tests are making this simple change so difficult to land. I'll review the test failures soon to figure out what to do.

**Comment by cjdeleon62** [comment]
@jridgewell no problem, thanks for helping out!

**Comment by jridgewell** [comment]
I pushed some updated tests to your branch, let's see if this fixes it.

**Comment by jridgewell** [comment]
And that did it. Thanks!

**Comment by cjdeleon62** [comment]
Thank you @jridgewell for working to get this merged in and helping out!

**Pull request merged**